### ISSUE TITLE
Is it possible to get time to first byte (TTFB)?

### ISSUE THREAD

**Issue by Kookabura** [scrapy/scrapy#2879]
Hello,
Is it possible to get time to first byte (TTFB) from request\response? I can get download_latency but this is a time including downloading. I've looked through middlewares but couldn't understand how to get TTFB.

**Comment by redapple** [comment]
@Kookabura , you probably need to patch Scrapy's [HTTP 1.1 download handler](https://github.com/scrapy/scrapy/blob/master/scrapy/core/downloader/handlers/http11.py#L302).

**Comment by Kookabura** [comment]
@redapple, thanks for pointing me. I've already tried it. Details are the below.
@kmike, glad to hear that you found it's useful.

I spent several days diving deeper into twisted and scrapy source to understand how it works and I didn't manage to get TTFB value that is near browser or curl values. As I understand it's not enough just patch self._cb_latency. Real TTFB might be got only from twisted.web.client. I've tried change `[_newclient.py](https://github.com/twisted/twisted/blob/trunk/src/twisted/web/_newclient.py#L347)` but unsuccessful.
But I'm not very good at python so I might miss something simple. Anyway I decided to stop experiments for some time as I a bit tired of unsuccessful coding.

**Comment by redapple** [comment]
[Wikipedia's page](https://en.wikipedia.org/wiki/Time_To_First_Byte) define TTFB as:
> This time is made up of the socket connection time, the time taken to send the HTTP request, and the time taken to get the first byte of the page. Although sometimes misunderstood as a post-DNS calculation, the original calculation of TTFB in networking always includes network latency in measuring the time it takes for a resource to begin loading.

As far as I understand, `_cb_latency` is called when all HTTP headers are received.
The latency in `request.meta['download_latency']` would therefore be _"the time taken to send the HTTP request, and the time taken to get all HTTP headers"_. The missing bit would be the time to get actual response body bytes.
Bytes from the response body may arrive later indeed, and these first bytes could probably be caught in `_ResponseReader`.
Am I missing something?

**Comment by Kookabura** [comment]
The next paragraph might be wrong. It's just my thoughts.
As scrapy asynchronous won't it lead to the situation when some requests wait before being handled? If that's true this waiting time will be added to the final result and the result might be wrong.

I think Wikipedia definition is right for general meaning. The situation is slightly different for web. If you look at TTFB in Chrome then you see something like [this](https://snag.gy/mBY8lZ.jpg). So I use Google explanation of TTFB. You can find it [here](https://developers.google.com/web/tools/chrome-devtools/network-performance/reference#timing). 

> Waiting (TTFB). The browser is waiting for the first byte of a response. TTFB stands for Time To First Byte. This timing includes 1 round trip of latency and the time the server took to prepare the response.

As far I understand it TTFB is the time after sending requests and before getting the first byte. Then TTFB should exclude the time on DNS-lookup and preparing the request.

**Comment by redapple** [comment]
@Kookabura , that level of detail indeed needs work at Twisted level (at least customizing the Agent, Request and HTTP protocol quite a bit)

**Comment by Kookabura** [comment]
Agree, but I didn't manage to develop the solution for Twisted yet. When I posted here I didn't know so many details as now. I was going to post question in Twisted repo  as well but they don't have Issues tab:).

**Comment by redapple** [comment]
For discussing this with Twisted, your safest bet is to use [their mailing list](https://twistedmatrix.com/trac/wiki/TwistedCommunity#MailLists).

**Comment by Kookabura** [comment]
@redapple Thank you, I'll give it a try.

**Comment by AdrianAtZyte** [comment]
https://github.com/scrapy/scrapy/pull/2906
